### ISSUE TITLE
Logistic regression fails in a pretty simple case

### ISSUE THREAD

**Issue by lendle** [JuliaStats/GLM.jl#105]
Set

```
y = [1.0,1,0,1,0,0,1,0,1,1,1,0,1,1,1,1,1,1,1,1,0,1,1,1,1,
     1,1,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,0,0,1,1,1,1,1,1,
     1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,
     1,1,1,1,1,1,1,1,1,1,1,1,1,1,0,1,1,1,1,1,1,1,1,1,1,
     1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,0,1,1,1,
     1,1,1,1,1,1,1,1,1,1,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,
     1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,0,1,1,1,
     1,1,1,1,1,1,1,1,1,1,0,1,1,1,1,0,1,1,1,1,1,1,1,1,1,
     0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,0,1,1,0,0,1,1,1,
     1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,0,1,1,1,1,1,1,1,1,1,
     1,1,1,1,1,1,1,1,1,1,1,1,1,0,1,1,1,1,1,1,1,1,1,1,1,
     1,1,1,1,1,1,1,1,1,1,1,1,1,1,0,1,1,1,1,1,1,1,1,1,1,
     1,1,1,1,1,1,1,0,1,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,
     0,0,1,1,1,1,1,1,1,1,1,0,1,1,1,1,1,1,1,1,1,1,1,1,1,
     1,1,1,1,1,1,1,1,1,1,1,1,1,0,1,1,1,1,1,1,1,1,1,1,0,
     1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,0,1,1,1,1,1,1,0,0,
     1,1,0,0,1,1,1,1,1,0,1,1,1,0,1,1,1,0,1,1,1,1,1,1,1,
     1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,
     1,1,1,1,1,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,
     1,1,0,1,1,1,1,1,1,0,1,1,1,1,1,1,1,1,0,1,1,1,1,1,1,
     1,1,0,1,1,1,1,1,1,1,1,1,1,1,1,1,0,1,1,1,1,1,1,0,1,
     1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,0,1,1,1,1,1,1,1,
     0,1,1,1,1,1,1,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,
     1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,
     1,1,1,1,1,1,1,0,0,1,1,1,1,1,1,1,1,1,1,1,1,0,1,1,1,
     1,1,1,1,1,1,1,1,1,1,1,1,0,1,1,1,1,1,1,1,1,0,1,1,0,
     1,1,1,0,1,0,1,1,0,1,1,1,1,1,1,1,1,1,1,0,1,0,1,1,1,
     1,1,1,1,1,1,1,1,0,1,1,1,1,1,1,1,1,1,0,1,1,1,1,1,1,
     1,0,1,1,1,1,1,1,0,1,1,1,1,1,1,1,1,0,1,1,1,1,1,0,1,
     1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,
     1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,
     1,1,1,1,0,1,1,1,1,1,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,
     1,1,1,1,1,1,1,1,1,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,
     1,1,1,1,1,1,1,1,1,1,1,1,1,1,0,1,1,1,1,1,1,1,1,0,1,
     1,1,0,1,1,1,1,1,1,1,1,1,0,1,1,1,1,1,1,1,1,1,1,1,1,
     0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,0,1,1,1,1,1,1,0,1,1,
     1,1,1,1,1,1,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,
     1,0,1,0,1,0,1,1,1,1,1,1,1,1,1,1,1,1,1,0,1,1,1,1,1,
     1,0,1,1,1,1,1,0,1,1,1,0,1,1,0,1,1,1,0,1,1,1,1,1,1,
     1,1,1,0,1,0,1,0,1,0,1,1,0,1,1,1,1,1,1,1,1,1,1,1,1]
```

``` jlcon
julia> mean(y)
0.911
```

I want to fit an intercept only logistic regression model with outcome `y` with a slightly stricter convergence criterion. `mean(y)` isn't too close to 0 or 1 so I wouldn't think there should be any numeric issues with this, but I haven't looked exactly at what the issue is.

``` jlcon
julia> using GLM

julia> fit(GeneralizedLinearModel, ones(1000, 1), y, Binomial(), convTol=1e-8)
ERROR: step-halving failed at beta0 = [2.3259063805025724]
 in error at error.jl:21
 in _fit at /Users/sam/.julia/v0.3/GLM/src/glmfit.jl:157
 in fit at /Users/sam/.julia/v0.3/GLM/src/glmfit.jl:176
 in fit at /Users/sam/.julia/v0.3/GLM/src/glmfit.jl:221
 in fit at /Users/sam/.julia/v0.3/GLM/src/glmfit.jl:206
```

If y is sorted everything is fine and the coefficient is what is expected: 

``` jlcon
julia> fit(GeneralizedLinearModel, ones(1000, 1), sort(y), Binomial(), convTol=1e-8)
GeneralizedLinearModel{GlmResp{Array{Float64,1},Binomial,LogitLink},DensePredChol{Float64}}:

Coefficients:
     Estimate Std.Error z value Pr(>|z|)
x1    2.32591  0.111057 20.9434   <1e-96

julia> p = mean(y)
0.911

julia> log(p/(1-p))
2.325906527527819
```

Any ideas as to what might be going on here?

**Comment by andreasnoack** [comment]
I can reproduce this on 0.3, but it appears to be gone on 0.4. Maybe the change of rounding mode has an effect here.

**Comment by simonster** [comment]
I managed to trigger this with another dataset on 0.4. Is the problem just that the deviance can appear to increase slightly after a Newton iteration because of floating point roundoff? Would it be reasonable to add an appropriate fudge factor to determine whether we need to do step-halving, or just consider the model to have converged if the absolute value of the relative change is less than the tolerance?

**Comment by lendle** [comment]
I'm seeing this again now with the normal distribution and identity link, and regressing on multiple covariates in addition to the intercept. Here's the data set if it's useful: http://pastebin.com/raw.php?i=S1vkQgX7. I'm regressing the 1st column on the rest + intercept.

**Comment by simonster** [comment]
Perhaps we should use the [R formula](https://stat.ethz.ch/R-manual/R-patched/library/stats/html/glm.control.html) to determine convergence, i.e., `abs(dev - devold)/(abs(dev) + 0.1) < convTol`?

**Comment by jotwin** [comment]
Got the same error

```
               _
   _       _ _(_)_     |  A fresh approach to technical computing
  (_)     | (_) (_)    |  Documentation: http://docs.julialang.org
   _ _   _| |_  __ _   |  Type "?help" for help.
  | | | | | | |/ _` |  |
  | | |_| | | | (_| |  |  Version 0.4.6-pre+4 (2016-03-19 15:00 UTC)
 _/ |\__'_|_|_|\__'_|  |  Commit 03c072d (9 days old release-0.4)
|__/                   |  x86_64-linux-gnu

julia> using GLM
  
julia> x=  [0 1; 1 1]
2x2 Array{Int64,2}:
 0  1
 1  1

julia> y = [11, 1]
2-element Array{Int64,1}:
 11
  1

julia> fit(GeneralizedLinearModel, x, y, Poisson())
ERROR: step-halving failed at beta0 = [-2.3978952727516103,2.3978952727983707]
 in error at ./error.jl:21
 in _fit! at /home/jakub/.julia/v0.4/GLM/src/glmfit.jl:152
 in fit at /home/jakub/.julia/v0.4/GLM/src/glmfit.jl:226
 in fit at /home/jakub/.julia/v0.4/GLM/src/glmfit.jl:229 (repeats 2 times)

julia> y = [11, 5]
2-element Array{Int64,1}:
 11
  5

julia> fit(GeneralizedLinearModel, x, y, Poisson())
GLM.GeneralizedLinearModel{GLM.GlmResp{Array{Float64,1},Distributions.Poisson,GLM.LogLink},GLM.DensePredChol{Float64,Base.LinAlg.Cholesky{Float64,Array{Float64,2}}}}:

Coefficients:
      Estimate Std.Error  z value Pr(>|z|)
x1   -0.788457   0.53936 -1.46184   0.1438
x2      2.3979  0.301511  7.95292   <1e-14


```

**Comment by phaverty** [comment]
Is this still an active issue? I get this "ERROR: step-halving failed at beta0" error in roughly 1 in 2000 runs (with different inputs). I have an application that does thousands of glm runs a day. I'd like to switch it from R to julia, but this issue is a blocker.

I'm using it like this:

f = @formula(Y ~ X)
mm = glm( f, df, Normal() )

Any help would be much appreciated!

**Comment by andreasnoack** [comment]
The are two questions here. The first is if we are reaching convergence in as many cases as, say, R or if we could do better. The other question is how to handle the convergence failures. We cannot guarantee that the algorithm converges so what should be done when it fails. Right now, we throw an exception. It seems that R just prints a warning. Would it be better to provide a flag that controls the error behavior?

**Comment by phaverty** [comment]
In a recent batch of work, GLM.jl gave this error on 8 / 110773 runs, which is a fair bit better than I originally reported. However, R's glm runs without error or warning and produces reasonable outputs. I think that GLM.jl should be able to be made to converge for these examples too.

A separate, but related issue is that GLM.jl gives an error for vectors with identical values, but R returns a fit with a NA for the coefficient. I'd prefer the NA myself.

julia> mydf = DataFrame(Y = [1,2,3,4,5], X = [1,1,1,1,1])
 5×2 DataFrames.DataFrame
 │ Row │ Y │ X │
 ├─────┼───┼───┤
 │ 1   │ 1 │ 1 │
 │ 2   │ 2 │ 1 │
 │ 3   │ 3 │ 1 │
 │ 4   │ 4 │ 1 │
 │ 5   │ 5 │ 1 │

!julia> glm(@formula(Y ~ X), mydf, Normal())
 ERROR: Base.LinAlg.PosDefException(2)
  in _chol!(::Array{Float64,2}, ::Type{UpperTriangular}) at ./linalg/cholesky.jl:55

> mydf = data.frame(Y = c(1,2,3,4,5), X = c(1,1,1,1,1))
 > glm(Y ~ X, gaussian, mydf)

 Call:  glm(formula = Y ~ X, family = gaussian, data = mydf)

 Coefficients:
 (Intercept)            X
           3           NA

 Degrees of Freedom: 4 Total (i.e. Null);  4 Residual
 Null Deviance:      10
 Residual Deviance: 10   AIC: 21.66

Just to be extra clear, the 8 errors mentioned above are not of this "Base.LinAlg.PosDefException" type.  Perhaps it would be better to have separate issues for "whether to error or just warn" and improved convergence.  I'd be happy to start the other issue.

**Comment by andreasnoack** [comment]
This is partly overlapping with the other issue. We should probably have a separate issue for discussion the general error behavior and ways to control it. The latest problem you show has the issue that the exception is thrown from a function we call instead of us throwing. I think we could get around this generally by using the pivoted Cholesky instead of the normal one. Then we might be able to zero out coefficients to get a basic solution similar to what (I think) R is doing. There might be some hidden issues with this approach. Maybe @dmbates has already considered this.

**Comment by dmbates** [comment]
The `R` solution is a bit more complicated than using the pivoted Cholesky.  R and S before it use a special pivoting method that moves a column to the right hand side when singularity is detected.  They both use a modified version of the Linpack (not LAPACK) QR decomposition to achieve this.  (Which, by the way, means that multi-threaded BLAS don't help on least squares calculations in R, much to people's dismay.)

The reason for not using a full pivoting is to keep columns corresponding to the same model term adjacent when possible, so that a sequential sum of squares decomposition is possible.  In a modern computing environment I don't think it is necessary to go to great lengths to make sum of squares decompositions easy so I would suggest using the Pivoted Cholesky.

I was going to suggest looking at a paper that Dirk Eddelbuettel and I wrote in J. Statist. Software about similar calculations in R using Rcpp and Eigen but then I looked at it again and realized that the C++ code obscures so much.  I am really glad to be programming numerical linear algebra in Julia, not in C++.
